### ISSUE TITLE
Question: Why are controllers continually reconstructed

### ISSUE THREAD

**Issue by dlgoodchild** [jeffijoe/awilix-koa#68]
Hi Jeff, before digging deeper, I am just looking to understand why a controller class appears to be constructed each time a route is requested/matched?

I.e. I create a class based controller and add to it a method annotated with `@route('/test')` and `@GET()`. Each time I request `GET /test` I will see that class being constructed. Is that the intention and if so, what is the behind-the-scenes reasoning for that?

Thanks.

**Comment by jeffijoe** [comment]
That's the intention, as each request creates a natural [scope](https://github.com/jeffijoe/awilix?tab=readme-ov-file#scoped-lifetime).

**Comment by dlgoodchild** [comment]
Would you be open to making that configurable, i.e. maybe default to using it, but allow an opt out of scoping controllers?

**Comment by jeffijoe** [comment]
I would be, yeah. Ideally that would be mirrored in the Express bindings, too.

**Comment by dlgoodchild** [comment]
OK that's great. I'll dig deeper then and have a look at testing an implementation. Once I've found what I need I'll summarise here for any feedback and open a PR for consideration.

**Comment by dlgoodchild** [comment]
I've been getting a deeper understanding of how it currently functions. I'm feeling like the options of importControllers/loadControllers could be more like:
```
options?: { glob: IOptions, scopeControllers: boolean = true } | IOptions
```
This would make it backwards compatible and potentially allow for adding more options in the future (should that ever be a thing). Any non-glob options would find their way into `_registerController`, where then the `scopeControllers` option would be passed into `makeInvoker` and drilled down to `makeResolverInvoker`.

If `scopeControllers = true` (default), then in `makeResolverInvoker` it will default to using `build`, otherwise it would register and resolve?

Thoughts/concerns?

**Comment by jeffijoe** [comment]
Pretty much, with a few minor tweaks:
1. I generally try to make opt-ins default to `false | undefined`, so we could name it `singleton`, that way we can rely on falsy checks rather than assigning a default value of `true`. Don't think we need the `Controllers` suffix, as that is implicit from context.
2. The controller should be registered with a unique symbol as the key, and lifetime `SINGLETON` to ensure strictness checks are performed.

**Comment by dlgoodchild** [comment]
Hi Jeff, have a look at this commit and let me know if there's anything you don't like:
https://github.com/jeffijoe/awilix-koa/compare/master...dlgoodchild:awilix-koa:master

Some comments that might help.
- I noticed that if you don't use `scopePerRequest` you get "can't call build on undefined" error, so I've added a helpful error for this.
- I don't have any use for creating container scopes per request, so I added an alternative to `scopePerRequest` in the form of `attachContainer`.
- Due to the design of using `ctx.state.container`, this means anything added to the container while using `scopePerRequest` is lost, so implementing `singleton` as above by itself wouldn't be enough.

The implementation considers all of the above and ensures there can be any combination.

Happy to take on any comments or feedback.

**Comment by jeffijoe** [comment]
Thanks! I like the addition of the error on undefined!

1. What's the implementation of `classOrFunctionSymbol` do? The name doesn't need to be unique in order for symbols to be unique, so that could possibly be simplified.
2. I'm not fond of the global fallback cache and visibility fields. Perhaps visibility is not necessary, since strict mode would throw an error when attempting to register a singleton in a scoped container?

Perhaps an alternative route here would be to avoid modifying the low-level invokers and instead perform the check and handling for singletons in `_registerController`? That way you could use the natural scope of that method to cache the instance of the controller and wouldn't need to register it at all.

If all of this ends up being too much trouble, it would be perfectly fine to just roll your own bindings that are tailored for your needs - the guts of the controller stuff is in `awilix-router-core` which the bindings packages just wrap per framework which you could lean on.

**Comment by dlgoodchild** [comment]
So, the `classOrFunctionSymbol` is simply creating a symbol name that we can can use each time there's an invocation against a particular controller, that then means we're not attempting to register the class or function again for each invocation. The contents of that method are effectively creating a hash using bitwise operations without needing any additional dependencies. We could actually just use it as a string key and drop the use of "symbol" to be honest.

I can simplify the code so that if you're using "scopedPerRequest", there's an automatic assumption that that you'd not want to use singleton controllers. This would reduce this change to simply having `attachContainer` and the `singleton` option?

How does that sound?

**Comment by dlgoodchild** [comment]
I've amended the commit accordingly which reduces the changes significantly and removes the use of symbol and the fallback array.

**Comment by jeffijoe** [comment]
Almost there - you don't need to register the singleton, you can simply `build` it and then store it in `resolved` and reuse when `singleton: true` - so that's even _less_ code 😄 

And as an aside, what I meant wrt `Symbol` was that symbols are innately unique, you wouldn't need a hash.

```js
const sym1 = new Symbol('hello')
const sym2 = new Symbol('hello')

console.log(sym1 === sym2) // false
```

But with these changes, no symbols or hash is needed.

**Comment by dlgoodchild** [comment]
OK no worries, simplified it further. My approach was to ensure that each function/class had a unique instance. In theory these changes mean I can use, say `controller` multiple times with the same function or class and get multiple instances. But I suppose the chances that someone might do that are slim.

I've amended the commit (linked earlier).

**Comment by jeffijoe** [comment]
Even in that case, the controller is bound to the route and so only one of them would be effective, registering the same controller twice would do nothing. Will take a look when I’m back!

**Comment by jeffijoe** [comment]
Looks good, final comments I can post on the PR in context when opened. 👍

**Comment by jeffijoe** [comment]
Implemented in #69
